### PR TITLE
Add unit tests for currency, parallel_scan, and mina_state

### PR DESCRIPTION
**Pull request by dkijania**
## Summary
- Add 736 lines of new unit tests across 5 files as Phase 1 of the test coverage improvement effort
- **currency**: scale, add/sub flagged, conversions, Fee_rate, Signed operations
- **parallel_scan**: empty state properties, free_space, sequence numbers, State.map, fold/hash
- **mina_state/ancestor**: Quickcheck verify rejects wrong ancestor/descendant
- **mina_state/local_state**: to_input determinism, gen distinctness
- **mina_state/snarked_ledger_state**: genesis state, to_input/to_field_elements determinism

Part of #18752

## Test plan
- [x] `dune build @check` passes
- [x] Code formatted with `make reformat-diff`
- [ ] `dune runtest src/lib/currency`
- [ ] `dune runtest src/lib/parallel_scan`
- [ ] `dune runtest src/lib/mina_state`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by dkijania** [comment]
!ci-build-me

**Comment by glyh** [comment]
Please move them to a separate file. We don't want a source file with 5000LoCs again.

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

File: src/lib/currency/test/test_currency.ml [hunk 1]
```diff
@@ -0,0 +1,384 @@
+open Core_kernel
+open Currency
+
+let%test_module "unchecked arithmetic" =
+  ( module struct
+    (* --- scale tests --- *)
+    let%test_unit "scale by zero returns zero" =
```
Comment (glyh) [inline src/lib/currency/test/test_currency.ml]: Maybe use quickcheck to do a probablistic property based test for all tests in this file?

File: src/lib/currency/test/test_currency.ml [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+    let%test_unit "sub self is zero" =
+      let v = Amount.of_uint64 (Unsigned.UInt64.of_int 999) in
+      [%test_eq: Amount.t option] (Amount.sub v v) (Some Amount.zero)
+
+    let%test_unit "add overflow returns none" =
+      [%test_eq: Amount.t option]
+        (Amount.add Amount.max_int Amount.max_int)
+        None
+
+    let%test_unit "sub underflow returns none" =
```
Comment (glyh) [inline src/lib/currency/test/test_currency.ml:21]: This feels bad as a behavior... That said, it's working code. Maybe document it in Currency source code if not already?

File: src/lib/currency/test/test_currency.ml (hunk 1 above)
Comment (dkijania) [reply src/lib/currency/test/test_currency.ml]: Ok, converted those suitabe. Only edge cases left

File: src/lib/currency/test/test_currency.ml (hunk 2 above)
Comment (dkijania) [reply src/lib/currency/test/test_currency.ml:21]: Documented

**Comment by dkijania** [comment]
!ci-build-me

File: src/lib/currency/test/test_currency.ml [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+          failwith "expected exception"
+
+    let%test_unit "of_mina_string_exn edge cases" =
+      (* whole number *)
+      let v1 = Amount.of_mina_string_exn "1" in
+      [%test_eq: Amount.t] v1 (Amount.of_mina_int_exn 1) ;
+      (* with decimals *)
+      let v2 = Amount.of_mina_string_exn "1.5" in
+      [%test_eq: Amount.t] v2
+        (Amount.of_uint64 (Unsigned.UInt64.of_int 1_500_000_000)) ;
```
Comment (glyh) [inline src/lib/currency/test/test_currency.ml:58]: It seems we should really rename `of_uint64` to `of_nanomina_uint64`, the current name is confusing.

File: src/lib/currency/test/test_currency.ml [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+      (* max precision *)
+      let v3 = Amount.of_mina_string_exn "0.000000001" in
+      [%test_eq: Amount.t] v3 (Amount.of_uint64 (Unsigned.UInt64.of_int 1))
+  end )
+
+let%test_module "signed operations" =
+  ( module struct
+    let mk n = Amount.of_uint64 (Unsigned.UInt64.of_int n)
```
Comment (glyh) [inline src/lib/currency/test/test_currency.ml:66]: I think this could be lifted to before, this pattern occurs pretty frequently, but I won't call it `mk`.

File: src/lib/currency/test/test_currency.ml [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+        Amount.Signed.create_preserve_zero_sign ~magnitude:Amount.zero
+          ~sgn:Sgn.Neg
+      in
+      assert (Amount.Signed.equal pos_zero neg_zero)
+
+    (* --- add_signed_flagged edge cases --- *)
+    let%test_unit "add_signed_flagged negative underflow" =
+      let base = mk 10 in
+      let delta = Amount.Signed.create ~magnitude:(mk 20) ~sgn:Sgn.Neg in
+      let _, `Overflow b = Amount.add_signed_flagged base delta in
```
Comment (glyh) [inline src/lib/currency/test/test_currency.ml:101]: This is technically an underflow. Maybe document on `add_signed_flagged` this weird corner case.

File: src/lib/currency/test/test_currency.ml [hunk 6]
```diff
@@ -0,0 +1,422 @@
...
+
+    let%test_unit "add_signed_flagged positive overflow" =
+      let delta = Amount.Signed.create ~magnitude:(mk 1) ~sgn:Sgn.Pos in
+      let _, `Overflow b = Amount.add_signed_flagged Amount.max_int delta in
+      assert b
+  end )
+
+let%test_module "balance operations" =
+  ( module struct
+    let mk n = Balance.of_uint64 (Unsigned.UInt64.of_int n)
```
Comment (glyh) [inline src/lib/currency/test/test_currency.ml:112]: Same as above, lift these helpers to above?

File: src/lib/currency/test/test_currency.ml [hunk 7]
```diff
@@ -0,0 +1,422 @@
...
+
+    let%test_unit "make creates valid rate" =
+      let r = Fee_rate.make (fee 100) 2 in
+      assert (Option.is_some r)
+
+    let%test_unit "make with zero weight" =
+      let r = Fee_rate.make (fee 0) 0 in
+      assert (Option.is_some r)
+
+    let%test_unit "to_uint64 on integer rate" =
```
Comment (glyh) [inline src/lib/currency/test/test_currency.ml:154]: I don't understand the operation here at all. fee rate is a production of fee and some weight. What does it even mean to convert it into an uint64?

There's no documentation either.

Comment (glyh) [reply src/lib/currency/test/test_currency.ml:154]: What is the "weight" here anyway?

File: src/lib/currency/test/test_currency.ml [hunk 8]
```diff
@@ -0,0 +1,422 @@
...
+    let%test_unit "sub self is zero" =
+      Quickcheck.test amount_gen ~trials:1000 ~f:(fun a ->
+          [%test_eq: Amount.t option] (Amount.sub a a) (Some Amount.zero) )
+
+    let%test_unit "add then sub roundtrip" =
+      Quickcheck.test (Quickcheck.Generator.tuple2 amount_gen amount_gen)
+        ~trials:1000 ~f:(fun (a, b) ->
+          match Amount.add a b with
+          | None ->
+              ()
```
Comment (glyh) [inline src/lib/currency/test/test_currency.ml:271]: Should probably filter out the none case with Generator.filter as well. Because the None case isn't testing the roundtrip.

File: src/lib/currency/test/test_currency.ml [hunk 9]
```diff
@@ -0,0 +1,422 @@
...
+          | Some c ->
+              [%test_eq: Amount.t option] (Amount.add c b) (Some a) )
+
+    let%test_unit "scale by 1 is identity" =
+      Quickcheck.test amount_gen ~trials:1000 ~f:(fun a ->
+          [%test_eq: Amount.t option] (Amount.scale a 1) (Some a) )
+
+    let%test_unit "scale by 0 is zero" =
+      Quickcheck.test amount_gen ~trials:1000 ~f:(fun a ->
+          [%test_eq: Amount.t option] (Amount.scale a 0) (Some Amount.zero) )
```
Comment (glyh) [inline src/lib/currency/test/test_currency.ml:290]: Looks like lots of PBTs here could just remove the hard-coded tests before. Just remove those hard-coded test if they're already covered in PBT?

**Comment by glyh** [comment]
I haven't done review on parallel_scan / mina_state at all. I suggest move them to a separate PR, or I'll need to read the implementation before actually decide if they're good or not.

File: src/lib/currency/test/test_currency.ml (hunk 3 above)
Comment (dkijania) [reply src/lib/currency/test/test_currency.ml]: Agree the name is unclear. Renaming `of_uint64` to `of_nanomina_uint64` (and the corresponding `to_uint64`) touches ~32 callsites across the repo and is a larger refactor than fits this PR's scope (test additions). I'll track it as a separate follow-up.

File: src/lib/currency/test/test_currency.ml (hunk 4 above)
Comment (dkijania) [reply src/lib/currency/test/test_currency.ml]: Lifted to file-top as `amount_of_int` / `balance_of_int` / `fee_of_int`. The local `mk` / `mk_amount` / `fee` helpers in the signed, balance, and fee_rate modules are gone.

File: src/lib/currency/test/test_currency.ml (hunk 5 above)
Comment (dkijania) [reply src/lib/currency/test/test_currency.ml:109]: Documented in `intf.ml` (both in module type `S` and in `Full.Amount`): the single ``Overflow` flag is set whenever the true result is out of range, regardless of direction — overflow on positive deltas and underflow on negative deltas are both reported via the same tag.

File: src/lib/currency/test/test_currency.ml (hunk 6 above)
Comment (dkijania) [reply src/lib/currency/test/test_currency.ml]: Done in the same change as the comment on line 66 — the `mk` / `mk_amount` helpers in this module are now the file-level `amount_of_int` / `balance_of_int`.

File: src/lib/currency/test/test_currency.ml (hunk 7 above)
Comment (dkijania) [reply src/lib/currency/test/test_currency.ml:160]: Documented the `Fee_rate` module in `intf.ml`. The new top-level comment explains it is fee-per-unit-of-weight (internally a rational [fee/weight]), and `to_uint64` now has a doc clarifying it returns `Some n` only when the rate is exactly the integer `n` (denominator = 1) and `None` for non-integer rates. `make` / `make_exn` got a similar refresh.

Comment (dkijania) [reply src/lib/currency/test/test_currency.ml:160]: "Weight" here is the unit of computational work the fee is being normalised against — e.g. a SNARK work unit. The new module-level doc on `Fee_rate` in `intf.ml` spells this out ("fee per unit of computational 'weight' (e.g. fee per unit of SNARK work)").

File: src/lib/currency/test/test_currency.ml (hunk 8 above)
Comment (dkijania) [reply src/lib/currency/test/test_currency.ml]: Switched to constructive generators built with `Generator.bind`: `add_defined_pairs_gen` picks any `a` and then picks `b` in `[0, max_int - a]`, so `Amount.add a b` is always `Some _`; `sub_defined_pairs_gen` does the analogous thing for subtraction. This is preferable to `Generator.filter` because the rejection rate would be high near the high end of the range — now every quickcheck trial actually exercises the roundtrip.

File: src/lib/currency/test/test_currency.ml (hunk 9 above)
Comment (dkijania) [reply src/lib/currency/test/test_currency.ml:309]: Removed the hand-written `Fee_rate` sexp roundtrip test (was `Fee_rate.make_exn (fee 100) 3` round-tripped through `sexp_of_t` / `t_of_sexp`) — it is fully subsumed by the property-based `Fee_rate sexp roundtrip` PBT in the `quickcheck properties` module. The other hand-written tests target specific edge cases that random generation can't reliably hit (max_int boundary overflow, `-1` for `of_nanomina_int` / `of_mina_int_exn`, exception paths, fixed-string parsing in `of_mina_string_exn`, and the zero-magnitude/zero-sign signed cases — `nonzero_amount_gen` excludes zero) so I left those in place as boundary anchors. Happy to prune more aggressively if you'd like.

**Comment by dkijania** [comment]
!ci-build-me
